src/librbd/image_copy/SnapshotCreateRequest.cc

```diff
@@ -0,0 +1,330 @@
+// vim: ts=8 sw=2 smarttab
+
+#include "SnapshotCreateRequest.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "cls/rbd/cls_rbd_types.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/ObjectMap.h"
+#include "librbd/Operations.h"
+#include "librbd/Utils.h"
+#include "osdc/Striper.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::image_copy::SnapshotCreateRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace image_copy {
+
+using librbd::util::create_context_callback;
+using librbd::util::create_rados_callback;
+
+template <typename I>
+SnapshotCreateRequest<I>::SnapshotCreateRequest(
+    I *dst_image_ctx, const std::string &snap_name,
+    const cls::rbd::SnapshotNamespace &snap_namespace,
+    uint64_t size, const librbd::ParentSpec &spec, uint64_t parent_overlap,
+    Context *on_finish)
+  : m_dst_image_ctx(dst_image_ctx), m_snap_name(snap_name),
+    m_snap_namespace(snap_namespace), m_size(size),
+    m_parent_spec(spec), m_parent_overlap(parent_overlap),
+    m_on_finish(on_finish), m_cct(dst_image_ctx->cct) {
+}
+
+template <typename I>
+void SnapshotCreateRequest<I>::send() {
+  send_set_size();
+}
+
+template <typename I>
+void SnapshotCreateRequest<I>::send_set_size() {
+  m_dst_image_ctx->snap_lock.get_read();
+  if (m_dst_image_ctx->size == m_size) {
+    m_dst_image_ctx->snap_lock.put_read();
+    send_remove_parent();
+    return;
+  }
+  m_dst_image_ctx->snap_lock.put_read();
+
+  ldout(m_cct, 20) << dendl;
+
+  // Change the image size on disk so that the snapshot picks up
+  // the expected size.  We can do this because the last snapshot
+  // we process is the sync snapshot which was created to match the
+  // image size. We also don't need to worry about trimming because
+  // we track the highest possible object number within the sync record
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::set_size(&op, m_size);
+
+  auto finish_op_ctx = start_dst_op();
+  if (finish_op_ctx == nullptr) {
+    lderr(m_cct) << "lost exclusive lock" << dendl;
+    finish(-EROFS);
+    return;
+  }
+
+  auto ctx = new FunctionContext([this, finish_op_ctx](int r) {
+      handle_set_size(r);
+      finish_op_ctx->complete(0);
+    });
+  librados::AioCompletion *comp = create_rados_callback(ctx);
+  int r = m_dst_image_ctx->md_ctx.aio_operate(m_dst_image_ctx->header_oid,
+                                                comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void SnapshotCreateRequest<I>::handle_set_size(int r) {
+  ldout(m_cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(m_cct) << "failed to update image size '" << m_snap_name << "': "
+                 << cpp_strerror(r) << dendl;
+    finish(r);
+    return;
+  }
+
+  {
+    // adjust in-memory image size now that it's updated on disk
```

Comment:
@trociny I believe It would be even better if we start comment with capital letter something as: 
// Adjust in-memory image size now that it's updated on disk